ast.cpp

```diff
@@ -44,47 +44,97 @@ object::object(){
     offSet=++counter*8;
 }
 
+CType::~CType() {}
+
 NodeBinOp::NodeBinOp(NodeExpr* l, NodeExpr* r, string p) {
     lhs = l;
     rhs = r;
     punct = p;
     lhs->parent = this;
     rhs->parent = this;
 }
-NodeAssign::NodeAssign(NodeExpr* _lhs, NodeExpr* _rhs) : NodeBinOp(_lhs, _rhs, "=") {}
 
-NodeDiv::NodeDiv(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "/") {}
+NodeAssign::NodeAssign(NodeExpr* _lhs, NodeExpr* _rhs) : NodeBinOp(_lhs, _rhs, "=") {
+    if(lhs->is_NodeId()) {
+        lhs->c_type = rhs->c_type;
+        var_map[((NodeId*)(lhs))->id]->c_type = lhs->c_type;
+    }
+    c_type = rhs->c_type;
+}
 
-NodeSub::NodeSub(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "-") {}
+NodeDiv::NodeDiv(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "/") {
+    c_type = new CIntType();
+}
 
-NodeAdd::NodeAdd(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "+") {}
+NodeSub::NodeSub(NodeExpr* l, NodeExpr* r) : NodeBinOp(l, r, "-") {
+    if(rhs->c_type->isIntType() && lhs->c_type->isPtrType()) {
```

Comment:
Set the c_type by handling all 4 cases of 
ptr - ptr 
ptr - num (this is the only case that's handled correctly, but we don't end up reading NodeSub's c_type in the codegen, only its children's types that handles all 4 cases. i think we need to check all 4 cases in both places, as in here and in codegen) 
num - ptr (not sure if this is suppposed to work, try with gcc or clang)
num - num (currently implemeneted in else block)